### PR TITLE
Add -t option to capput to use tombs for deletion

### PR DESCRIPTION
**Pull request by beckerg**
Signed-off-by: Greg Becker <gbecker@micron.com>

### CODE REVIEW

File: lib/cn/csched_sp3.c [hunk 1]
```diff
@@ -323,7 +323,7 @@ scale2dbl(u64 samp)
 static inline uint
 samp_est(struct cn_samp_stats *s, uint scale)
 {
-    return scale * safe_div(s->i_alen + s->l_alen, s->l_good);
+    return scale * safe_div((s->i_alen / 4) + s->l_alen, s->l_good);
```
Comment (tristan957) [inline lib/cn/csched_sp3.c:326]: what is the 4 here?

File: lib/cn/csched_sp3.c [hunk 2]
```diff
@@ -932,37 +931,45 @@ sp3_dirty_node_locked(struct sp3 *sp, struct cn_tree_node *tn)
...
+         * "split_pct" capacity to ensure it gets split or compacted.
+         *
+         * Node splits are rare, but when a node is ready to split it should
+         * be done as soon as possible.  Hence, splits prevent other "big"
+         * compaction jobs from starting which could otherwise indefinitely
+         * starve a split.
          */
         if (tn->tn_ns.ns_pcap >= sp->thresh.lcomp_split_pct && jobs < 1) {
             const uint64_t weight = ((uint64_t)tn->tn_ns.ns_pcap << 32) | keys;
+            uint64_t split_keys = (uint64_t)sp->thresh.lcomp_split_keys << 22;
```
Comment (tristan957) [inline lib/cn/csched_sp3.c:947]: Where do these shift values come from?

File: tools/capput/capput.c [hunk 3]
```diff
@@ -157,6 +159,77 @@ pdel(void *arg)
...
+
+    txn = hse_kvdb_txn_alloc(targ->kvdb);
+    if (!txn)
+        fatal(ENOMEM, "Failed to allocate resources for txn");
+
+    pthread_setname_np(pthread_self(), __func__);
+
+    while (!killthreads) {
+        char key[sizeof(uint64_t) * 2];
+        uint64_t *p; /* prefix */
```
Comment (tristan957) [inline tools/capput/capput.c:178]: Just call the variables prefix and suffix and then remove the comment :)

File: lib/cn/csched_sp3.c (hunk 1 above)
Comment (beckerg) [reply lib/cn/csched_sp3.c:326]: That just reduces the impact of the root node on space amp by 4x.  Alex is going to get rid of this altogether, but this gets us closer to how we expect to be operating in the near future.  Basically it reduced space-amp thrashing which can trigger unnecessary gc.

File: lib/cn/csched_sp3.c [hunk 4]
```diff
@@ -932,37 +931,45 @@ sp3_dirty_node_locked(struct sp3 *sp, struct cn_tree_node *tn)
...
 
             sp3_node_insert(sp, spn, wtype_split, weight);
-            splitting = true;
+
+            if ((uint)tn->tn_ns.ns_pcap * 2 > (uint)sp->thresh.lcomp_split_pct * 3 ||
+                keys_uniq * 3 > split_keys * 2) {
+
```
[the remark this answers is not in the document]
Comment (beckerg) [reply lib/cn/csched_sp3.c:953]: No good answer here, but this is much easier to read the messy predicates when they stand out...

File: lib/cn/csched_sp3.c (hunk 2 above)
Comment (beckerg) [reply lib/cn/csched_sp3.c:947]: See the comments for the variables and related defines.  Actually, the threshold structure is unnecessarily compacted which causes stuff like this and casting to larger type.  If we uncompacted it we could eliminate all this silly mess.  I'll open a ticket..

File: lib/cn/csched_sp3.c (hunk 1 above)
Comment (tristan957) [reply lib/cn/csched_sp3.c:326]: Sounds good

File: tools/capput/capput.c (hunk 3 above)
Comment (beckerg) [reply tools/capput/capput.c:178]: Copied from the other function.  The presence of globals named "pfx" and "sfx" and similar cause a lot of confusion and I suspect is what drove the use of p and s.  Feel free to create a ticket and rename everything, I think it an exercise of diminishing return at this point...

File: lib/cn/csched_sp3.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (beckerg) [reply lib/cn/csched_sp3.c:953]: Hmmm.  it's indented correctly to the left of the open paren (i.e., at the same scope of the first predicate).
Once I fix the thresh var sizes the casts can go away and it should fit all on one line :-)

File: lib/cn/csched_sp3.c (hunk 2 above)
Comment (beckerg) [reply lib/cn/csched_sp3.c:947]: https://jira.micron.com/jira/browse/NFSE-5332
This has been irritating me for quite a while, thanks for pointing it out!

File: tools/capput/capput.c (hunk 3 above)
Comment (beckerg) [reply tools/capput/capput.c:178]: Hahaha!  Sigh...

**Review by gsramdasi — APPROVED** [review]

**Review by nabeelmmd — APPROVED** [review]

File: lib/cn/csched_sp3.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (alexttx) [reply lib/cn/csched_sp3.c:953]: Late to the party, but when the condition spans multiple lines I like to force an open brace and put it on a new line.
```
if ((uint)tn->tn_ns.ns_pcap * 2 > (uint)sp->thresh.lcomp_split_pct * 3 ||
    keys_uniq * 3 > split_keys * 2)
{
    ...
}
```

**Pull request merged**